### PR TITLE
General fixes

### PR DESCRIPTION
**Pull request by SirCmpwn**

### CODE REVIEW

File: osu.Framework/Allocation/DependencyContainer.cs [hunk 1]
```diff
@@ -105,8 +105,10 @@ private void Register(Type type, bool lazy)
         /// <summary>
         /// Caches an instance of a type. This instance will be returned each time you Get<T>.
         /// </summary>
-        public T Cache<T>(T instance = null, bool lazy = false) where T : class
+        public T Cache<T>(T instance = null, bool overwrite = false, bool lazy = false) where T : class
         {
+            if (!overwrite && cache.ContainsKey(typeof(T)))
+                throw new InvalidOperationException(@"We have already cached one of these");
```
Comment (Tom94) [inline osu.Framework/Allocation/DependencyContainer.cs]: Let's use `Assert` here? As far as I can tell caching something twice is a programmer error. We shouldn't expect it to ever happen in a correct program.

File: osu.Framework/BaseGame.cs [hunk 2]
```diff
@@ -112,7 +112,7 @@ private void load()
             Audio = Dependencies.Cache(new AudioManager(
                 new NamespacedResourceStore<byte[]>(Resources, @"Tracks"),
                 new NamespacedResourceStore<byte[]>(Resources, @"Samples")));
-            Dependencies.Cache(Audio);
+            Dependencies.Cache(Audio, overwrite: true);
```
Comment (peppy) [inline osu.Framework/BaseGame.cs]: we don't use named parameters anywhere else (and it's not even required here). any reason for this?

Comment (SirCmpwn) [inline osu.Framework/BaseGame.cs]: explicit > implicit

Comment (peppy) [inline osu.Framework/BaseGame.cs]: so should we go through and use it everywhere? even when it's not necessary due to order of parameters..?

Comment (SirCmpwn) [inline osu.Framework/BaseGame.cs]: I think it should be left to developer discretion. If the author of the code thinks it makes the code more understandable in each case, then so be it. In this case `true` really doesn't have any obvious meaning without looking up the method.

Comment (peppy) [inline osu.Framework/BaseGame.cs]: i don't think it looks more understandable. it just confuses me that we're suddenly using it in one place where we never have before. intellisense already shows what the argument's name is, so i don't like this syntax.

Comment (SirCmpwn) [inline osu.Framework/BaseGame.cs]: Intellisense isn't always there, like when reading this diff on github. Does this really confuse you?

Comment (peppy) [inline osu.Framework/BaseGame.cs]: i don't like it and would be against introducing named parameters in the project as a whole

Comment (SirCmpwn) [inline osu.Framework/BaseGame.cs]: Alright, w/e

File: osu.Framework/Allocation/DependencyContainer.cs [hunk 3]
```diff
@@ -105,8 +105,9 @@ private void Register(Type type, bool lazy)
         /// <summary>
         /// Caches an instance of a type. This instance will be returned each time you Get<T>.
         /// </summary>
-        public T Cache<T>(T instance = null, bool lazy = false) where T : class
+        public T Cache<T>(T instance = null, bool overwrite = false, bool lazy = false) where T : class
         {
+            Debug.Assert(overwrite || cache.ContainsKey(typeof(T)), @"We have already cached one of these");
```
Comment (peppy) [inline osu.Framework/Allocation/DependencyContainer.cs]: shouldn't the second condition be negated?

File: osu.Framework/BaseGame.cs [hunk 4]
```diff
@@ -112,7 +112,7 @@ private void load()
             Audio = Dependencies.Cache(new AudioManager(
                 new NamespacedResourceStore<byte[]>(Resources, @"Tracks"),
                 new NamespacedResourceStore<byte[]>(Resources, @"Samples")));
-            Dependencies.Cache(Audio);
+            Dependencies.Cache(Audio, true);
```
Comment (peppy) [inline osu.Framework/BaseGame.cs]: why is this required?

Comment (SirCmpwn) [inline osu.Framework/BaseGame.cs]: Because it overwrites an AudioManager that was cached by Host iirc

Comment (peppy) [inline osu.Framework/BaseGame.cs]: that sounds like a bug?
